packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -70,34 +80,115 @@ impl<K> Default for DependencyMap<K> {
+                    identifier.clone(),
+                    (new_resolve_depth, new_interval.clone()),
+                );
+                DependencyStatus::Unresolved(new_resolve_depth, new_interval)
             }
             RawEntryMut::Occupied(entry) => {
-                if entry.into_mut().update(resolved_depth) {
-                    DependencyStatus::Unresolved
-                } else {
+                let (current_depths, current_interval) = entry.into_mut();
+                let old_interval = current_interval.clone();
+                *current_interval = current_interval.clone().merge(new_interval.clone());
+
+                if current_depths.update(new_resolve_depth) {
+                    // It does not matter if the interval is updated or not, as it has to be
+                    // resolved for the full interval anyway.
+                    DependencyStatus::Unresolved(*current_depths, current_interval.clone())
+                } else if old_interval.contains_interval(&new_interval) {
+                    tracing::info!("{old_interval:?} contains {new_interval:?}");
+                    // The dependency is already resolved for the required interval
+                    // old: [-----)
+                    // new:  [---)
                     DependencyStatus::Resolved
+                } else if new_interval.contains_interval(&old_interval)
+                    || new_interval.is_adjacent_to(&old_interval)
+                {
+                    // The dependency is already resolved, but not for the required interval or the
+                    // intervals are adjacent so we only have to resolve the new interval
+                    // old:  [---)   |  [---)
+                    // new: [-----)  |      [---)
+                    DependencyStatus::Unresolved(*current_depths, new_interval)
+                } else if old_interval.overlaps(&new_interval) {
+                    // The time intervals either overlap or are disjoint, so only have to resolve
+                    // the difference of the two intervals.
+                    // old:     [-----)
+                    // new:       [-------)
+                    // resolve:       [---)
+                    let difference = new_interval.difference(old_interval);
+                    // The intervals do overlap and the current interval is not a subset of
+                    // the required interval, so the difference must be a single interval
+                    debug_assert_eq!(difference.len(), 1, "difference must be a single interval");
+
+                    DependencyStatus::Unresolved(
+                        *current_depths,
+                        difference
+                            .into_iter()
+                            .next()
+                            .expect("difference must be a single interval"),
+                    )
+                } else {
+                    // The time intervals are disjoint. Ideally, we only would resolve the new
```

Comment:
The intervals can be disjoint as well above, right? If we're going to bother to start the comment with this line I think we should try and be comprehensive. Is the correct statement that they're disjoint and non-adjacent?

Reply:
Actually, the comment above was wrong. I fixed the comment above in d2e49fd417e2928f50fb1a454ad09996db9f44a4 and added, that the intervals are not adjacent either in b02de604b0598abcc5cadc1a0042149d12d51956.

Reply:
Thank you :)